lib/reek/smells/smell_detector.rb

```diff
@@ -38,10 +38,19 @@ def contexts
       end
 
       def run_for(context)
-        return [] unless enabled_for?(context)
         return [] if exception?(context)
 
-        sniff(context)
+        smells = sniff(context)
+
+        if enabled_for?(context)
```

Comment:
Then I would move the second if .. else into a separate method, and even move the call to `sniff` inside the conditional:

```ruby
if enabled_for? context
  sniff(context)
else
  check_for_unneeded_suppression(context)
end
```

or even (going back to the original code layout):

```ruby
return check_for_unneeded_suppression(context) unless enabled_for? context
sniff(context)
```